presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/ShardIterator.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.raptor.metadata;
+
+import com.facebook.presto.raptor.util.CloseableIterator;
+import com.facebook.presto.spi.PrestoException;
+import com.google.common.collect.AbstractIterator;
+import com.google.common.collect.ImmutableSet;
+
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.UUID;
+
+import static com.facebook.presto.raptor.RaptorErrorCode.RAPTOR_ERROR;
+import static com.facebook.presto.raptor.util.UuidUtil.uuidFromBytes;
+
+final class ShardIterator
+        extends AbstractIterator<ShardNodes>
+        implements CloseableIterator<ShardNodes>
+{
+    private final Connection connection;
+    private final PreparedStatement statement;
+    private final ResultSet resultSet;
+
+    private boolean done;
+    private UUID currentShardId;
+    private ImmutableSet.Builder<String> nodes = ImmutableSet.builder();
+
+    public ShardIterator(String sql, ShardPredicate predicate, Connection connection)
+    {
+        this.connection = connection;
+        try {
+            statement = connection.prepareStatement(sql);
+            predicate.bind(statement);
+            resultSet = statement.executeQuery();
+        }
+        catch (SQLException e) {
+            close();
+            throw new PrestoException(RAPTOR_ERROR, e);
+        }
+    }
+
+    @Override
+    protected ShardNodes computeNext()
+    {
+        try {
+            return compute();
+        }
+        catch (SQLException e) {
+            throw new PrestoException(RAPTOR_ERROR, e);
```

Comment:
This is a `Closeable`, so callers are always required to close it. We only need to close in the constructor because the object isn't constructed yet.

As an example, the `read()` method of `FileInputStream` doesn't close itself when an `IOException` occurs.